sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala

```diff
@@ -157,10 +193,11 @@ object UnsupportedOperationChecker extends Logging {
     // Disallow multiple streaming aggregations
     val aggregates = collectStreamingAggregates(plan)
 
-    if (aggregates.size > 1) {
+    if (aggregates.size > 1 && outputMode != InternalOutputModes.Append) {
```

Comment:
I think it's not only multiple aggregations. Reasoning streaming aggregation followed by time window join with update mode would be very tricky. I'd say we should simply disallow multiple stateful operators with update and complete mode at all.
(There could be some combinations where they still make sense for update/complete mode, but it doesn't seem to be trivial to reason and list up allowlist or blocklist.)

cc. @alex-balikov

Reply:
Absolutely. I agree that we should allow multiple stateful ops only in append mode. The other modes are not implemented.